Multi namespace

bshephar opened the pull request.

bshephar commented on api/v1beta1/openstackprovisionserver.go:
```diff
@@ -11,7 +12,6 @@ import (
 	ctx context.Context,
 	c goClient.Client,
-	instance *OpenStackProvisionServer,
 ) (map[string]int32, error) {
```
If we implement this as proposed, then we can probably also simplify this data structure. We could just create a slice of port integers rather than a map.

abays commented:
Is `OpenStackProvisionServer` currently not working for port assignment at the cluster scope?  We had implemented [1] to allow multi-namespace functionality, which was verified by Gil in his environment.  Just curious if you found something broken, which I'm guessing you did.

[1] https://github.com/openstack-k8s-operators/openstack-baremetal-operator/pull/230

hjensas reviewed (commented):
I tested this, and it did not work. See discussion on slack https://redhat-internal.slack.com/archives/CQXJFGMK6/p1745493175279419?thread_ts=1745488572.085949&cid=CQXJFGMK6

bshephar commented:
> I tested this, and it did not work. See discussion on slack https://redhat-internal.slack.com/archives/CQXJFGMK6/p1745493175279419?thread_ts=1745488572.085949&cid=CQXJFGMK6

Thanks, marking it as draft while we continue to explore the problem.

bshephar commented:
I think I see where I went wrong, pushing an update to this PR and I'll test it.

bshephar commented:
> Is `OpenStackProvisionServer` currently not working for port assignment at the cluster scope? We had implemented [1] to allow multi-namespace functionality, which was verified by Gil in his environment. Just curious if you found something broken, which I'm guessing you did.
> 
> [1] #230

I just updated this PR to use the `OpenStackProvisionServer` name + namespace as the key in the map. That leaves the original logic untouched, which I think should work. wdyt @abays ?

bshephar commented:
The failing functional test here actually validates that the change works:
```
  2025-04-29T09:08:36+10:00     ERROR   Reconciler error        {"controller": "openstackprovisionserver", "controllerGroup": "baremetal.openstack.org", "controllerKind": "OpenStackProvisionServer", "OpenStackProvisionServer": {"nam
e":"edpm-compute-baremetalset-provisionserver","namespace":"bddf94f8-31e6-46c4-93af-f57e692a16db"}, "namespace": "bddf94f8-31e6-46c4-93af-f57e692a16db", "name": "edpm-compute-baremetalset-provisionserver", "reconcileID": "66c2437d-7
022-40c8-826b-c697d5f6f757", "error": "admission webhook \"vopenstackprovisionserver.kb.io\" denied the request: port 6191 is already in use by another OpenStackProvisionServer"}
```

Because we do the whole new namespace for each test thing, so we end up creating the provisionServer instances in separate namespaces with the `BeforeEach()`. 

If I update the webhook to provide the name of the conflicting instance, we can see that it's two instances in different namespaces:
```
  2025-04-29T09:14:17+10:00     ERROR   Reconciler error        {"controller": "openstackprovisionserver", "controllerGroup": "baremetal.openstack.org", "controllerKind": "OpenStackProvisionServer", "OpenStackProvisionServer": {"nam
e":"edpm-compute-baremetalset-provisionserver","namespace":"9fefe8e4-803d-43eb-bd83-4e90028b08da"}, "namespace": "9fefe8e4-803d-43eb-bd83-4e90028b08da", "name": "edpm-compute-baremetalset-provisionserver", "reconcileID": "5e04d5c3-8
91f-4816-9908-867979e7e4e6", "error": "admission webhook \"vopenstackprovisionserver.kb.io\" denied the request: port 6190 is already in use by another OpenStackProvisionServer: edpm-compute-baremetalset-provisionserver-9fefe8e4-803
d-43eb-bd83-4e90028b08da"}
```

abays commented:
> The failing functional test here actually validates that the change works:
> 
> ```
>   2025-04-29T09:08:36+10:00     ERROR   Reconciler error        {"controller": "openstackprovisionserver", "controllerGroup": "baremetal.openstack.org", "controllerKind": "OpenStackProvisionServer", "OpenStackProvisionServer": {"nam
> e":"edpm-compute-baremetalset-provisionserver","namespace":"bddf94f8-31e6-46c4-93af-f57e692a16db"}, "namespace": "bddf94f8-31e6-46c4-93af-f57e692a16db", "name": "edpm-compute-baremetalset-provisionserver", "reconcileID": "66c2437d-7
> 022-40c8-826b-c697d5f6f757", "error": "admission webhook \"vopenstackprovisionserver.kb.io\" denied the request: port 6191 is already in use by another OpenStackProvisionServer"}
> ```
> 
> Because we do the whole new namespace for each test thing, so we end up creating the provisionServer instances in separate namespaces with the `BeforeEach()`.
> 
> If I update the webhook to provide the name of the conflicting instance, we can see that it's two instances in different namespaces:
> 
> ```
>   2025-04-29T09:14:17+10:00     ERROR   Reconciler error        {"controller": "openstackprovisionserver", "controllerGroup": "baremetal.openstack.org", "controllerKind": "OpenStackProvisionServer", "OpenStackProvisionServer": {"nam
> e":"edpm-compute-baremetalset-provisionserver","namespace":"9fefe8e4-803d-43eb-bd83-4e90028b08da"}, "namespace": "9fefe8e4-803d-43eb-bd83-4e90028b08da", "name": "edpm-compute-baremetalset-provisionserver", "reconcileID": "5e04d5c3-8
> 91f-4816-9908-867979e7e4e6", "error": "admission webhook \"vopenstackprovisionserver.kb.io\" denied the request: port 6190 is already in use by another OpenStackProvisionServer: edpm-compute-baremetalset-provisionserver-9fefe8e4-803
> d-43eb-bd83-4e90028b08da"}
> ```

Testing this in a live environment, I see this error cycling in the operator logs:

```
2025-05-01T08:40:03Z    ERROR   Metadate update failed  {"controller": "openstackprovisionserver", "controllerGroup": "baremetal.openstack.org", "controllerKind": "OpenStackProvisionServer", "OpenStackProvisionServer": {"name":"openstack","namespace":"openstack"}, "namespace": "openstack", "name": "openstack", "reconcileID": "85b93428-66d8-4ffb-81bd-fd05cfa6ac16", "error": "admission webhook \"vopenstackprovisionserver.kb.io\" denied the request: port 6191 is already in use by another OpenStackProvisionServer"}
```

But the initial port assignment looks good:

```
$ oc get osprovserver -o json -A | jq -r '.items[].spec.port'
6191
6190
```

Because of the error in the operator, however, the `OpenStackProvisionServer`s are stuck:

```
$ oc get osprovserver -A
NAMESPACE    NAME        STATUS   MESSAGE
openstack    openstack            
openstack2   openstack
```

bshephar commented:
> > The failing functional test here actually validates that the change works:
> > ```
> >   2025-04-29T09:08:36+10:00     ERROR   Reconciler error        {"controller": "openstackprovisionserver", "controllerGroup": "baremetal.openstack.org", "controllerKind": "OpenStackProvisionServer", "OpenStackProvisionServer": {"nam
> > e":"edpm-compute-baremetalset-provisionserver","namespace":"bddf94f8-31e6-46c4-93af-f57e692a16db"}, "namespace": "bddf94f8-31e6-46c4-93af-f57e692a16db", "name": "edpm-compute-baremetalset-provisionserver", "reconcileID": "66c2437d-7
> > 022-40c8-826b-c697d5f6f757", "error": "admission webhook \"vopenstackprovisionserver.kb.io\" denied the request: port 6191 is already in use by another OpenStackProvisionServer"}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Because we do the whole new namespace for each test thing, so we end up creating the provisionServer instances in separate namespaces with the `BeforeEach()`.
> > If I update the webhook to provide the name of the conflicting instance, we can see that it's two instances in different namespaces:
> > ```
> >   2025-04-29T09:14:17+10:00     ERROR   Reconciler error        {"controller": "openstackprovisionserver", "controllerGroup": "baremetal.openstack.org", "controllerKind": "OpenStackProvisionServer", "OpenStackProvisionServer": {"nam
> > e":"edpm-compute-baremetalset-provisionserver","namespace":"9fefe8e4-803d-43eb-bd83-4e90028b08da"}, "namespace": "9fefe8e4-803d-43eb-bd83-4e90028b08da", "name": "edpm-compute-baremetalset-provisionserver", "reconcileID": "5e04d5c3-8
> > 91f-4816-9908-867979e7e4e6", "error": "admission webhook \"vopenstackprovisionserver.kb.io\" denied the request: port 6190 is already in use by another OpenStackProvisionServer: edpm-compute-baremetalset-provisionserver-9fefe8e4-803
> > d-43eb-bd83-4e90028b08da"}
> > ```
> 
> Testing this in a live environment, I see this error cycling in the operator logs:
> 
> ```
> 2025-05-01T08:40:03Z    ERROR   Metadate update failed  {"controller": "openstackprovisionserver", "controllerGroup": "baremetal.openstack.org", "controllerKind": "OpenStackProvisionServer", "OpenStackProvisionServer": {"name":"openstack","namespace":"openstack"}, "namespace": "openstack", "name": "openstack", "reconcileID": "85b93428-66d8-4ffb-81bd-fd05cfa6ac16", "error": "admission webhook \"vopenstackprovisionserver.kb.io\" denied the request: port 6191 is already in use by another OpenStackProvisionServer"}
> ```
> 
> But the initial port assignment looks good:
> 
> ```
> $ oc get osprovserver -o json -A | jq -r '.items[].spec.port'
> 6191
> 6190
> ```
> 
> Because of the error in the operator, however, the `OpenStackProvisionServer`s are stuck:
> 
> ```
> $ oc get osprovserver -A
> NAMESPACE    NAME        STATUS   MESSAGE
> openstack    openstack            
> openstack2   openstack
> ```

The functional tests I added here now reproduce that issue. I'll keep digging.

bshephar commented:
Ok, the issue was that the webhook checked for the `r.Name` instead of the `r.Name + "-" + r.Namespace` like we're setting in the existingPorts map

rabi commented on api/v1beta1/openstackprovisionserver.go:
```diff
@@ -25,7 +25,7 @@ func GetExistingProvServerPorts(
 
 	for _, provServer := range provServerList.Items {
-		found[provServer.Name] = provServer.Spec.Port
+		found[fmt.Sprintf("%s-%s", provServer.Name, provServer.Namespace)] = provServer.Spec.Port
```
Should we use `namespacedName` type as key instead?
```
		namespacedName := types.NamespacedName{
			Name: provServer.Name, Namespace: provServer.Namespace }
		found[namespacedName.String()] = provServer.Spec.Port
```

bshephar replied:
I had a look at this, so it would be handy if I already had a `types.NamespacedName` on the `instance`. But since I don't, I would need to import the apimachinery types here in this module which currently doesn't import it, then I need to build the type before referencing it. I'm not sure the extra step is adding a lot of value?

To avoid confusion, the latest update here stops referring to it as a NamespacedName

rabi commented on tests/functional/openstackbaremetalset_controller_test.go:
```diff
@@ -174,21 +190,9 @@ var _ = Describe("BaremetalSet Test", func() {
-			osImageDir := "/usr/local/apache2/htdocs"
 
 			provServer := GetProvisionServer(baremetalSetName)
-			spec := baremetalv1.OpenStackProvisionServerSpec{
```
Why remove assert for all these fields?

rabi reviewed (approved):
Importing apimachinery packages again here won't have any impact on executable size or performance (loaded once).  Anyway, let's merge this to fix the issue, we can rework this later.

rabi commented:
/lgtm
